tests/unit/cpu_profiler_alloc_test.cc

```diff
@@ -0,0 +1,67 @@
+/* This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2023 ScyllaDB Ltd.
+ */
+
+#include <chrono>
+#define BOOST_TEST_MODULE core
+
+#include <boost/test/unit_test.hpp>
+
+#include <string_view>
+
+#include <seastar/core/internal/cpu_profiler.hh>
+#include <seastar/core/memory.hh>
+#include <seastar/core/smp.hh>
+
+template <typename Func>
+SEASTAR_CONCEPT(requires requires(Func fn) { fn(); })
+void check_function_allocation(const char *name, size_t expected_allocs,
+                               Func f) {
+  auto before = seastar::memory::stats();
+  f();
+  auto after = seastar::memory::stats();
+
+  BOOST_TEST_INFO("After function: " << name);
+  BOOST_REQUIRE_EQUAL(expected_allocs, after.mallocs() - before.mallocs());
```

Comment:
also check before.mallocs() > 0 (will require release mode)? Otherwise we might be not testing anything.

Reply:
Not sure what we'd gain by checking if `before.mallocs() > 0` as we're only concerned with testing if the `after - before` difference is equal to some amount(0 for this case). A concern could be that the compiler may reorder `f()` outside of the assignments of before and after. In which case we wouldn't be testing anything. Was this your concern?